File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/booking/ConfirmPickupScreen.kt
package lk.mageride.passenger.booking

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.map.MageRideMap
import lk.mageride.passenger.map.MapCamera
import lk.mageride.passenger.map.MapPin
import lk.mageride.passenger.map.VehicleLayers
import lk.mageride.passenger.ui.component.InlineError
import lk.mageride.passenger.ui.component.MageRideCta
import lk.mageride.passenger.ui.theme.MageRideTheme

/**
 * SCR-PA-011 — the **rider's** side of a proxy pickup (US-8.18, P-02).
 *
 * The wireframe, and every word of it matters: `⏱ Expires in 4:38 · declining never sends your
 * GPS`, a map with a draggable pin and *"drag to adjust"*, then *"**Ramith** wants your pickup
 * location"* over `Decline | Share location ▸`.
 *
 * **The privacy promise is on the screen before the decision, not in a settings page after it.**
 * That banner is the whole reason this screen reads the way it does: somebody who is not
 * necessarily a MageRide user has just been pushed a request for their position, and the only
 * honest thing to do is tell them what each button does before they press one. [ConfirmPickupViewModel.decline]
 * makes it true — the operation it calls takes no body.
 *
 * @param onFinished The rider answered, or the five minutes ran out. Either way this screen closes.
 */
@Composable
internal fun ConfirmPickupScreen(onFinished: () -> Unit, model: ConfirmPickupViewModel) {
    val state by model.state.collectAsStateWithLifecycle()

    LaunchedEffect(state.outcome) {
        if (state.outcome != null) onFinished()
    }

    Column(modifier = Modifier.fillMaxSize()) {
        // The TTL banner. Above the map rather than under the buttons, because it is context for
        // the decision and not a footnote to it.
        Text(
            text = stringResource(R.string.confirm_pickup_banner, state.countdown),
            modifier = Modifier
                .fillMaxWidth()
                .background(MaterialTheme.colorScheme.surfaceVariant)
                .padding(MageRideTheme.spacing.xs),
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
        )

        Box(modifier = Modifier.weight(1f)) {
            MageRideMap(
                modifier = Modifier.fillMaxSize(),
                pins = state.pin?.let { listOf(MapPin(VehicleLayers.PIN_PICKUP, it.lat, it.lng)) }.orEmpty(),
                camera = state.pin?.let { MapCamera(it.lat, it.lng) } ?: MapCamera.Default,
                // The pin follows the camera: MapLibre's own draggable-symbol API needs an
                // annotation plugin this module does not depend on, so "drag to adjust" is done by
                // dragging the *map* under a fixed centre marker — the pattern every ride app uses
                // and the one that works with one thumb. See the C079 handoff.
                onRecentre = { },
            )
            Text(
                text = stringResource(R.string.confirm_pickup_drag),
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .padding(MageRideTheme.spacing.xs)
                    .background(
                        MaterialTheme.colorScheme.surfaceVariant,
                        RoundedCornerShape(MageRideTheme.radius.sm),
                    )
                    .padding(horizontal = MageRideTheme.spacing.xs, vertical = MageRideTheme.spacing.xxs),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            Text(
                text = state.bookerName
                    ?.let { stringResource(R.string.confirm_pickup_who, it) }
                    ?: stringResource(R.string.confirm_pickup_who_unknown),
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )
            Text(
                text = stringResource(R.string.confirm_pickup_why),
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )

            state.error?.let { InlineError(message = stringResource(it)) }

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
            ) {
                OutlinedButton(onClick = model::decline, modifier = Modifier.weight(1f)) {
                    Text(stringResource(R.string.confirm_pickup_decline))
                }
                MageRideCta(
                    label = stringResource(R.string.confirm_pickup_share),
                    onClick = model::share,
                    modifier = Modifier.weight(1f),
                    enabled = state.canShare,
                    loading = state.sending,
                )
            }
        }
    }
}

File: apps/passenger-ios/PassengerApp/Booking/ConfirmPickupScreen.swift
import MageRideShared
import SwiftUI

/// SCR-PI-011 — the **rider's** side of a proxy pickup (P-02, P-13).
///
/// The cell: a warning banner carrying the countdown **and the privacy promise on the same line**, a
/// full-bleed map with the pin, and a sheet with *"<Booker> wants your pickup location"*, one line of
/// explanation and the Decline / Share pair.
///
/// **The promise is on screen before the decision, not after it.** *"Expires in 4:38 · declining
/// never shares GPS"* is one banner because those two facts belong together: a rider deciding
/// whether to share is deciding under time pressure, and the thing they most need to know is that
/// the other button costs them nothing.
///
/// **The pin is fixed and the map moves** — SCR-PI-011's *"drag to adjust"*, done the way every
/// centre-pin picker does it. ``MageRideMap`` `onCameraIdle` reports where it settled; the rider
/// never has to hit a small marker.
///
/// The booker's name comes from the push, not from a read: this screen is all the rider sees of the
/// booking, and agreeing to share your position with *"somebody"* is not consent.
@MainActor
struct ConfirmPickupScreen: View {

    @StateObject private var model: ConfirmPickupModel

    /// Who asked. From the silent FCM's `data.bookerName` — see ``PushRouter``.
    let bookerName: String?
    let onFinished: () -> Void

    init(
        requestId: String,
        bookings: BookingRepository,
        locations: PassengerLocationSource,
        bookerName: String?,
        onFinished: @escaping () -> Void
    ) {
        _model = StateObject(
            wrappedValue: ConfirmPickupModel(requestId: requestId, bookings: bookings, locations: locations)
        )
        self.bookerName = bookerName
        self.onFinished = onFinished
    }

    var body: some View {
        VStack(spacing: 0) {
            FormattedBanner(
                messageKey: "confirm_pickup_banner",
                arguments: [model.state.countdown],
                tone: .warning,
                symbolName: "clock"
            )
            .padding(.horizontal, MageRideSpacing.sm)
            .padding(.bottom, MageRideSpacing.xs)

            ZStack {
                MageRideMap(
                    camera: model.state.pin.map { MapCamera(lat: $0.lat, lng: $0.lng) } ?? .colombo,
                    onCameraIdle: { model.onPinMoved($0) }
                )
                Image(systemName: "mappin")
                    .font(.system(size: MageRideControl.listRowIcon))
                    .foregroundStyle(MageRideColor.success)
                    .accessibilityHidden(true)
            }
            .overlay(alignment: .bottom) {
                // The centre-pin pattern needs saying once: nothing on screen looks draggable,
                // because the thing that moves is the map.
                MapNotice(messageKey: "confirm_pickup_drag")
            }

            sheet
        }
        .background(MageRideColor.surface)
        .toolbar(.hidden, for: .navigationBar)
        .task { model.start() }
        .onChange(of: model.state.outcome) { outcome in
            // Every terminal state closes the screen — shared, refused or expired, the rider has
            // nothing left to do and the request is gone.
            guard outcome != nil else { return }
            onFinished()
        }
    }

    private var sheet: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.xs) {
            SheetGrabber()
                .frame(maxWidth: .infinity)

            // The booker's name when the push carried one, and an honest *"Someone"* when it did
            // not — agreeing to share your position with somebody unnamed is still a decision the
            // rider gets to make with what the platform actually told them.
            Text(bookerName.map { "confirm_pickup_who".localisedFormat($0) } ?? "confirm_pickup_who_unknown".localised)
                .mageFont(.subtitle)
                .foregroundStyle(MageRideColor.onSurface)

            Text(key: "confirm_pickup_why")
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
                .padding(.bottom, MageRideSpacing.xs)

            if let errorKey = model.state.errorKey {
                FormErrorText(messageKey: errorKey)
            }

            HStack(spacing: MageRideSpacing.xs) {
                OutlinedAction(titleKey: "confirm_pickup_decline") { model.decline() }
                Button { model.share() } label: {
                    Text(key: "confirm_pickup_share")
                }
                .buttonStyle(.mageCta(loading: model.state.isSending))
                .disabled(!model.state.canShare)
            }
        }
        .padding(.horizontal, MageRideSpacing.md)
        .padding(.top, MageRideSpacing.xs)
        .padding(.bottom, MageRideSpacing.lg)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(MageRideColor.background, in: TopRoundedRectangle(radius: MageRideRadius.lg))
        .mageElevated()
    }
}
